ext/opentelemetry-ext-pymemcache/src/opentelemetry/ext/pymemcache/__init__.py

```diff
@@ -0,0 +1,176 @@
+# limitations under the License.
+
+"""
+
+Usage
+-----
+
+The OpenTelemetry ``pymemcache`` integration traces pymemcache client operations
+
+Usage
+-----
+
+.. code-block:: python
+
+    from opentelemetry import trace
+    from opentelemetry.sdk.trace import TracerProvider
+    from opentelemetry.ext.pymemcache import PymemcacheInstrumentor
+    trace.set_tracer_provider(TracerProvider())
+    PymemcacheInstrumentor().instrument()
+    from pymemcache.client.base import Client
+    client = Client(('localhost', 11211))
+    client.set('some_key', 'some_value')
+
+API
+---
+"""
+# pylint: disable=no-value-for-parameter
+
+import logging
+
+import pymemcache
+from wrapt import ObjectProxy
+from wrapt import wrap_function_wrapper as _wrap
+
+from opentelemetry.ext.pymemcache.util import _get_address_attributes
+from opentelemetry.ext.pymemcache.version import __version__
+from opentelemetry.instrumentation.instrumentor import BaseInstrumentor
+from opentelemetry.trace import SpanKind, get_tracer
+from opentelemetry.trace.status import Status, StatusCanonicalCode
+
+logger = logging.getLogger(__name__)
+
+_DEFAULT_SERVICE = "memcached"
+_RAWCMD = "db.statement"
+_CMD = "memcached.command"
+COMMANDS = [
+    "set",
+    "set_many",
+    "add",
+    "replace",
+    "append",
+    "prepend",
+    "cas",
+    "get",
+    "get_many",
+    "gets",
+    "gets_many",
+    "delete",
+    "delete_many",
+    "incr",
+    "decr",
+    "touch",
+    "stats",
+    "version",
+    "flush_all",
+    "quit",
+    "set_multi",
+    "get_multi",
+]
+
+
+def _set_connection_attributes(span, instance):
+    for key, value in _get_address_attributes(instance).items():
+        span.set_attribute(key, value)
+
+
+def _with_tracer_wrapper(func):
+    """Helper for providing tracer for wrapper functions.
+    """
+
+    def _with_tracer(tracer, cmd):
+        def wrapper(wrapped, instance, args, kwargs):
+            # prevent double wrapping
+            if hasattr(wrapped, "__wrapped__"):
+                return wrapped(*args, **kwargs)
+
+            return func(tracer, cmd, wrapped, instance, args, kwargs)
+
+        return wrapper
+
+    return _with_tracer
+
+
+@_with_tracer_wrapper
+def _wrap_cmd(tracer, cmd, wrapped, instance, args, kwargs):
+    with tracer.start_as_current_span(
+        _CMD, kind=SpanKind.INTERNAL, attributes={}
+    ) as span:
+        try:
+            if not args:
+                vals = ""
+            else:
+                vals = _get_query_string(args[0])
+
+            query = "{}{}{}".format(cmd, " " if vals else "", vals)
+            span.set_attribute(_RAWCMD, query)
+
+            _set_connection_attributes(span, instance)
+        except Exception as ex:  # pylint: disable=broad-except
+            logger.warning(
+                "Failed to set attributes for pymemcache span %s", str(ex)
+            )
+
+        return wrapped(*args, **kwargs)
+
+
+def _get_query_string(arg):
```

Comment:
I'm wondering if some of these methods defined here could go into `util.py` and leave only the wrapping, instrumenting code in `__init__.py`.